File: src/leetcode/python/easy/BinaryTreeLevelOrderTraversalII_107.py
from collections import deque


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinaryTreeLevelOrderTraversalII_107(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        result = []
        queue = deque([root])
        size = len(queue)
        while size != 0:
            level = []
            for i in range(size):
                node = queue.popleft()
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            result.append(level)
            size = len(queue)
        return result[::-1]

File: src/leetcode/java/BinaryTreeLevelOrderTraversalII_107.java
package leetcode.java.easy;

import java.util.LinkedList;
import java.util.List;

public class BinaryTreeLevelOrderTraversalII_107 {
  public List<List<Integer>> levelOrderBottom(TreeNode root) {
    List<List<Integer>> result = new LinkedList<>();
    if (root == null) return result;
    LinkedList<TreeNode> queue = new LinkedList<>();
    queue.add(root);
    TreeNode node = null;
    int size = 0;
    while ((size = queue.size()) != 0) {
      LinkedList<Integer> level = new LinkedList<>();
      for (int i = 0; i < size; i++) {
        node = queue.removeFirst();
        level.add(node.val);
        if (node.left != null) queue.add(node.left);
        if (node.right != null) queue.add(node.right);
      }
      ((LinkedList) result).addFirst(level);
    }
    return result;
  }

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;
  TreeNode(int x) { val = x; }
}
}